src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java

```diff
@@ -38,69 +39,117 @@
+      + "<span style=\"font-style: normal; font-size: 125%; font-weight: 750;\">"
+      + "202 Accepted</span>"
+      + " - The response body will contain an object of type"
+      + " <span style=\"font-style: normal; font-size: 125%; font-weight: 750;\">"
+      + DatasetIdDto.SWAGGER_MODEL_NAME + "</span>.";
 
   private static final String MESSAGE_FOR_RETRIEVE_DATASET = ""
-          + "<span style=\"font-style: normal; font-size: 125%; font-weight: 750;\">"
-          + "200 OK</span>"
-          + " - The response body will contain an object of type"
-          + " <span style=\"font-style: normal; font-size: 125%; font-weight: 750;\">"
-          + ProgressInfoDto.PROGRESS_SWAGGER_MODEL_NAME + "</span>.";
+      + "<span style=\"font-style: normal; font-size: 125%; font-weight: 750;\">"
+      + "200 OK</span>"
+      + " - The response body will contain an object of type"
+      + " <span style=\"font-style: normal; font-size: 125%; font-weight: 750;\">"
+      + ProgressInfoDto.PROGRESS_SWAGGER_MODEL_NAME + "</span>.";
 
   private static final Pattern namePattern = Pattern.compile("[a-zA-Z0-9_-]+");
 
   @Value("${sandbox.dataset.max-size}")
   private int maxRecords;
 
-  private final ZipService zipService;
+  private final HarvestService harvestService;
   private final DatasetService datasetService;
   private final DatasetReportService reportService;
 
-  public DatasetController(ZipService zipService,
+
+  public DatasetController(HarvestService harvestService,
       DatasetService datasetService,
       DatasetReportService reportService) {
-    this.zipService = zipService;
+    this.harvestService = harvestService;
     this.datasetService = datasetService;
     this.reportService = reportService;
   }
 
-  @ApiOperation("Process the given dataset")
+  @ApiOperation("Process the given dataset by HTTP providing a file")
+  @ApiResponses({
+      @ApiResponse(code = 202, message = MESSAGE_FOR_PROCESS_DATASET, response = Object.class)
+  })
+  @PostMapping(value = "dataset/{name}/harvestByFile", produces = APPLICATION_JSON_VALUE)
+  @ResponseStatus(HttpStatus.ACCEPTED)
+  public DatasetIdDto harvestDatasetFromFile(
+      @ApiParam(value = "name of the dataset", required = true) @PathVariable(value = "name") String datasetName,
+      @ApiParam(value = "country of the dataset", required = true, defaultValue = "Netherlands") @RequestParam Country country,
+      @ApiParam(value = "language of the dataset", required = true, defaultValue = "nl") @RequestParam Language language,
+      @ApiParam(value = "dataset records uploaded in a zip file", required = true) @RequestParam MultipartFile dataset) {
+    checkArgument(namePattern.matcher(datasetName).matches(),
+        "dataset name can only include letters, numbers, _ or - characters");
+
+    List<ByteArrayInputStream> records = harvestService.harvest(dataset);
+
+    checkArgument(records.size() < maxRecords,
+        "Amount of records can not be more than " + maxRecords);
```

Comment:
The checkArguments done here are the same or very as the other API calls. Could you work around this to avoid code duplication?
You can, for example, create a private method to check the input values and each API call can use it

Reply:
I'll just leave this as it is. If you want to discuss this offline. Let me know

Reply:
It's always preferable to improve the quality of our code, and I believe that it is possible to include the `checkArguments` and the `harvest` function into one private method, give it a try, just as we mentioned in our call ;) 
If it gets too complicated, then just leave it be.